### ISSUE TITLE
Behavior of softmax when axis is set

### ISSUE THREAD

**Issue by maxbartel** [onnx/onnx-mlir#1876]
Hi!

When converting softmax from onnx to the kernel dialect without setting the `axis` attribute 
```mlir
func.func @test_softmax(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
  %1 = "onnx.Softmax"(%arg0) : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
  func.return %1 : tensor<13x21x3xf32>
}
```
the loops are as expected, so:
```mlir
// Outerloop
krnl.iterate(%0#0, %0#1) with (%0#0 -> %arg1 = 0 to 13, %0#1 -> %arg2 = 0 to 21)
// Innerloops 
krnl.iterate(%2) with (%2 -> %arg3 = 0 to 3)
```

But when setting the `axis` attribute, instead of switching just the axis used for the innerloop against the other axis, all axes after the attribute are moved from the outerloop to the innerloop. For example:
```mlir
func.func @test_softmax(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
  %1 = "onnx.Softmax"(%arg0) {axis = 0 : si64} : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
  func.return %1 : tensor<13x21x3xf32>
}
```
leads to 
```mlir
// Outerloop
// Not there anymore
// Innerloops 
krnl.iterate(%2#0, %2#1, %2#2) with (%2#0 -> %arg1 = 0 to 13, %2#1 -> %arg2 = 0 to 21, %2#2 -> %arg3 = 0 to 3)
```

Another example: 
From
```mlir
func.func @test_softmax(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
  %2 = "onnx.Softmax"(%arg0) {axis = 1 : si64} : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
  func.return %2 : tensor<13x21x3xf32>
}
```
to
```mlir
// Outerloop
krnl.iterate(%0) with (%0 -> %arg1 = 0 to 13)
// Innerloops 
krnl.iterate(%2#0, %2#1) with (%2#0 -> %arg2 = 0 to 21, %2#1 -> %arg3 = 0 to 3)
```

Is this intended behavior? If so, what is the reasoning behind it? In the [PyTorch Softmax](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html) for example, just the axis would be switched out. 

Thank you for your help!

**Comment by tungld** [comment]
Hi @maxbartel ONNXSoftmax has several versions. Axis in version 13 means it is the dimension to do softmax, while in versions < 13 it means all dimensions after the axis are coreced. For more info: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Softmax

By default, the lowering of softmax to krnl assumes version 11. If you would like to see the behaviour of version 13, set the attribute `onnx_opset = 13 : si64`, for example `%0 = "onnx.Softmax"(%arg0) {axis = 1: si64, onnx_opset = 13 : si64}`.

**Comment by maxbartel** [comment]
Hi @tungld,
Thank you for your help! This cleared things up for me.

But I have a follow-up question:
In Version 13 the default value of axis is -1, while in version <13 it was 1. If the user does not set `onnx_opset = 13 : si64`, he will assume to have an older version of softmax. However, the axis value if not set explicitly is -1, the default of the newer version. Is it also expected that the user then always sets the axis value to 1?

**Comment by ljfitz** [comment]
Hi @tungld - if the semantics are opset dependent, shouldn't this be addressed in the `version_dict` of `gen_onnx_mlir.py` as happens for `Clip` and a few others?

**Comment by tungld** [comment]
> In Version 13 the default value of axis is -1, while in version <13 it was 1

Good catch @maxbartel, I think we should change to using `version_dict` as @ljfitz mentioned to avoid the confusion here. Really appriciate it if one of you is interested in making this change.

**Comment by ljfitz** [comment]
Thanks @tungld for the direction - we'll open a PR to make the change.
